### ISSUE TITLE
Cities fails to use CPU cache when looping plots

### ISSUE THREAD

**Issue by Nightinggale** [Religion-And-Revolution-Extended/Mod#103]
When CvCity loops the plots within range, it does so in a spiral, meaning it goes from closest to furthest away. This could make sense for looping, which stops looping when the first usable candidate is found. However that's not what happens. Instead it goes through every single plot within range.

The problem with this approach is that CvMap stores the plots linear in memory, meaning they appear like the beam on a CRT monitor. This means looping plots should be done:

```cpp
for (y=top to bottom)
    for (x=left to right)
```

The actual implementation could be different, but if this read order is used, the CPU cache will be more likely to contain the next read plot, hence reducing memory I/O latency.

Most likely this affects all code, which loops 3x3 and 5x5 plots. We should investigate how widespread this issue is, but most likely cities are the worst in terms of performance. The AI loops colony plots a lot of times to figure out which professions each unit should have.

**Comment by abmpicoli** [comment]
Would you mind place an example? Also, even if CvCity loops the plots in spiral , and uses all plots, there are situations where this is desirable: the tiles at the corners of the city radius have the most chance of being occupied by neighbor cities when they expand, while tiles close to the city plot are "more safe". So, if the AI chooses between two equal plots, the one closest to home should be better...

**Comment by Nightinggale** [comment]
There are two parts of the code. One is loop order, the other is AI priority. I agree that the AI priority should stay the same and not start to favor the far north-western corner.

The vanilla approach is to store int arrays in CvGlobals and then loop the arrays to follow predefined plot order (see below). Those arrays are used by the functions plotCity, plotCityXY(2 overloaded functions) in CvGameCoreUtils.cpp.

One example of plot looping is CvCityAI::AI_buildingValue(). It loops all plots to ask isWater() to each when figuring out how useful it will be to build docks. There are plenty of such examples if you start to look for them.

Improving regarding memory I/O is good, but we should also consider which coding strategy we would want to use here. For instance should CvCity hold an array of plot pointers instead of looking all over the code/memory to figure out how to loop.

(part of) the code in question from CvGlobals::init()
```cpp
	int aiCityPlotPriority[NUM_CITY_PLOTS] =
	{
		0,
		1, 2, 1, 2, 1, 2, 1, 2,
#ifndef ONE_PLOT_CITY_RADIUS
		3, 4, 4, 3, 4, 4, 3, 4, 4, 3, 4, 4,	// R&R, ray, 2 Plot Radius
		4, 4, 4, 4, // R&R, ray, 2 Plot Radius
#endif
	};

	int aaiXYCityPlot[CITY_PLOTS_DIAMETER][CITY_PLOTS_DIAMETER] =
	{
#ifdef ONE_PLOT_CITY_RADIUS
		{6, 7, 8,},
		{5, 0, 1,},
		{4, 3, 2,},
#else
		{23, 17, 18, 19, 24,},	// R&R, ray, 2 Plot Radius - START
		{16, 6, 7, 8, 20,},
		{15, 5, 0, 1, 9,},
		{14, 4, 3, 2, 10,},
		{22, 13, 12, 11, 21,}	// R&R, ray, 2 Plot Radius - END
#endif
	};
```

**Comment by ShadesOT** [comment]
part of the problem is, that it is all calculations with discreet values. i am sure the algorithm can be optimized.

if the map generation algorithm was several sin/cos and other functions, though, it just would need the coordinates and the code would almost instantly know the plot values of an 5x5 area for anything (eg if it is good for some building or what yield to grow, etc ... ).

additionally all that needs to be stored is some seed values for the map generation functions.

topography functions would also allow the AI to fast determine where to build cities.

in short, everything would be faster, smaller (and better ;-).

**Comment by ShadesOT** [comment]
algorithms like eg this

https://en.m.wikipedia.org/wiki/Diamond-square_algorithm

**Comment by Nightinggale** [comment]
Whatever we do, we have to remember this affects calculations done parallel in network games without communication, meaning the game just assumes identical results. As a result, we can't use code, which has a risk of floating point rounding error or anything else of that nature as it will risk the AI not doing the same on all computers, hence desync.
